src/bootstrap/dist.rs

```diff
@@ -2009,17 +2009,30 @@ impl Step for RustDev {
     fn run(self, builder: &Builder<'_>) -> Option<GeneratedTarball> {
         let target = self.target;
 
+        let mut tarball = Tarball::new(builder, "rust-dev", &target.triple);
+
+        let bootstrap_outdir = &builder.bootstrap_out;
+        for file in &[
+            "bootstrap",
+            "libbootstrap.rlib",
+            "llvm-config-wrapper",
+            "rustc",
+            "rustdoc",
+            "sccache-plus-cl",
+        ] {
+            tarball.add_file(bootstrap_outdir.join(file), "bin", 0o755);
+        }
+
+        tarball.set_overlay(OverlayKind::LLVM);
+
         /* run only if llvm-config isn't used */
         if let Some(config) = builder.config.target_config.get(&target) {
             if let Some(ref _s) = config.llvm_config {
-                builder.info(&format!("Skipping RustDev ({}): external LLVM", target));
-                return None;
+                builder.info(&format!("Skipping llvm ({}): external LLVM", target));
+                return Some(tarball.generate());
```

Comment:
I think this isn't the right behavior. With your changes now, we could end up with a tarball that has just bootstrap and not LLVM. In CI that doesn't matter, because we always build LLVM from source, but if a distro is running `x.py dist` from source, they'll end up with the rust-dev component even though they don't actually need it.

I think the proper fix is to add a `default_condition` to `should_run` above that doesn't run this step by default when using external LLVM. That will still let you test this locally while building LLVM, and won't break CI, but also avoids building rust-dev without LLVM unless you specifically ask for it.

Reply:
hmm how i can access to the `target_config` from `should_run`?

Reply:
nvm ShouldRun has a builder field

Reply:
We chatted a bit more about this [on Zulip](https://rust-lang.zulipchat.com/#narrow/stream/122652-new-members/topic/distribute.20bootstrap.20in.20CI.20.2398483/near/287448035). This is non-trivial to get working because to set a `default_condition` we need to know which platform we're targeting. But bootstrap is target-independent (wants to be `ONLY_HOSTS = true`) while LLVM is target-dependent; plus if we turn this off by default, I'd worry that we'd miss adding it in one of the builders.

@Mark-Simulacrum how do you feel about separating the LLVM component from the bootstrap component? I know the original motivation for calling this `rust-dev` was for bundling them, but that seems like more effort than it's worth ... plus it means you have to download and unpack LLVM even if you're only planning to use bootstrap and compile llvm from source. I think having a separate `llvm-nightly` and `bootstrap-nightly` component makes more sense, and is easier to maintain.

Reply:
I'm OK with separating them; I don't think I would suffix with -nightly. Maybe -internal could work well; the indicator we want to give is "not really intended for consumption outside of bootstrap itself".

Reply:
That said, I'm not sure that we need to? It seems to me that it'd be totally fine to only produce this component when enabled (perhaps a new config.toml option for ease of passing it to all the dist builders).

Reply:
@Mark-Simulacrum so, my original concern was that someone would run `x dist` intending to get a toolchain and get `rust-dev` packaged along with it. Right now that doesn't usually happen because most distros are using a system LLVM instead of building from source. After this change, they'll get rust-dev even with system llvm (it will have just bootstrap, not the pre-compiled LLVM).

If you think that's ok, we don't need to change anything and can keep this as a default step.

Reply:
That's also fine; I'm not really worried about us producing a few extra artifacts. If we want to avoid it, I think the right way is to add a config.toml option (internal-artifacts or whatever name) to gate this step.